### PR TITLE
[export] update dynamic shapes section

### PR DESCRIPTION
**Pull request by pianpwk**
## Description
Updates the export dynamic shapes section to be more up-to-date with the typical user workflow

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### CODE REVIEW

File: intermediate_source/torch_export_tutorial.py [hunk 1]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+        self.l = torch.nn.Linear(5, 3)
+
+    def forward(
+        self,
+        w: torch.Tensor,  # [6, 5]
+        x: torch.Tensor,  # [4]
+        y: torch.Tensor,  # [8, 4]
+        z: torch.Tensor,  # [32]
+    ):
+        x0 = x + y  # output shape: [8, 4]
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:323]: do you need to say "output shape"? IMO it's better to establish the convention in a sentence before the examples ("we'll mention the shapes of local variables in inline comments" or some such thing).

File: intermediate_source/torch_export_tutorial.py [hunk 2]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+        x1 = self.l(w)  # [6, 3]
+        x2 = x0.flatten()  # [32]
+        x3 = x2 + z  # [32]
+        return x1, x3
+
+######################################################################
+# By default, ``torch.export`` produces a static program. One clear consequence of this is that at runtime,
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:330]: remove "clear"

File: intermediate_source/torch_export_tutorial.py [hunk 3]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+}
+ep = export(model, (w, x, y, z), dynamic_shapes=dynamic_shapes)
 
-    def forward(self, x, y):
-        return torch.nn.functional.relu(self.lin(x + y), inplace=True)
+######################################################################
+# Before we look at the program that's produced, let's understand what specifying ``dynamic_shapes`` entails,
+# and how that interacts with export. For every input dimension where a ``Dim`` object is specified, a symbol is
+# allocated, taking on a range of ``[2, inf]`` (why not ``[0, inf]`` or ``[1, inf]``? we'll explain later in the
+# 0/1 specialization section).
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:361]: IMO it's worth linking to https://pytorch.org/docs/main/export.programming_model.html#basics-of-symbolic-shapes when you talk about "a symbol is allocated."

File: intermediate_source/torch_export_tutorial.py [hunk 4]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+    def forward(
+        self,
+        w: torch.Tensor,  # [s0, s1]
+        x: torch.Tensor,  # [s2]
+        y: torch.Tensor,  # [s3, s4]
+        z: torch.Tensor,  # [s5]
+    ):
+        x0 = x + y  # guard: s2 == s4
+        x1 = self.l(w)  # guard: s1 == 5
+        x2 = x0.flatten()
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:386]: mention no guard added here; also may be worth adding the shapes like above, only now symbolic...

File: intermediate_source/torch_export_tutorial.py [hunk 5]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
-try:
-    exported_dynamic_shapes_example1.module()(torch.randn(8, 20, 2))
-except Exception:
-    tb.print_exc()
-
-try:
-    exported_dynamic_shapes_example1.module()(torch.randn(8, 8, 3))
-except Exception:
-    tb.print_exc()
+print(ep)
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:409]: add the result of print? it would be cool for people to relate the symbols they see for nodes in the graph here (to get used to notation)...

File: intermediate_source/torch_export_tutorial.py [hunk 6]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
-print(exported_dynamic_shapes_example2.module()(torch.randn(2, 16), torch.randn(16, 4)))
+# This spits out quite a handful, even with this simple toy model. But looking through the logs we can see the lines relevant
+# to what we described above; e.g. the allocation of symbols:
 
-try:
-    exported_dynamic_shapes_example2.module()(torch.randn(4, 8), torch.randn(4, 2))
-except Exception:
-    tb.print_exc()
+"""
+I1210 16:20:19.720000 3417744 torch/fx/experimental/symbolic_shapes.py:4404] [1/0] create_symbol s0 = 6 for L['w'].size()[0] [2, int_oo] (_dynamo/variables/builder.py:2841 in <lambda>), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s0" or to suppress this message run with TORCHDYNAMO_EXTENDED_ADVICE="0"
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:432]: perhaps remove the prefix of each line until the file name? could say "<snip>" and add a few words after "spits out quite a handful" to explain you're doing this for readability.

File: intermediate_source/torch_export_tutorial.py [hunk 7]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
 ######################################################################
-# We can actually use ``torch.export`` to guide us as to which ``dynamic_shapes`` constraints
-# are necessary. We can do this by relaxing all constraints (recall that if we
-# do not provide constraints for a dimension, the default behavior is to constrain
-# to the exact shape value of the example input) and letting ``torch.export``
-# error out.
-
-inp4 = torch.randn(8, 16)
-inp5 = torch.randn(16, 32)
+# Next to the ``[guard added]`` messages, we also see the responsible user lines of code - luckily here the model is simple enough.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:451]: I like that you're calling out `[guard added]`, so likewise above call out `create_symbol` and ask the user to search based on that...

File: intermediate_source/torch_export_tutorial.py [hunk 8]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+# In many real-world cases it's not so straightforward: high-level torch operations can have complicated fake-kernel implementations
+# or operator decompositions that complicate where and what guards are emitted. In such cases the best way to dig deeper and investigate
+# is to follow the logs' suggestion, and re-run with environment variable ``TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="..."``, to further
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:454]: does this help? if so, maybe paste what happens; otherwise omit?

File: intermediate_source/torch_export_tutorial.py [hunk 9]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+# attribute the guard of interest.
+#
+# ``Dim.AUTO`` is just one of the available options for interacting with ``dynamic_shapes``; as of writing this 2 other options are available:
+# ``Dim.DYNAMIC``, and ``Dim.STATIC``. ``Dim.STATIC`` simply marks a dimension static, while ``Dim.DYNAMIC`` is similar to ``Dim.AUTO`` in all
+# ways except one: it raises an error when specializing to a constant; designed to maintain dynamism. See for example what happens when a
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:459]: `; designed ...` -> maybe `; this is designed ...`

File: intermediate_source/torch_export_tutorial.py [hunk 10]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
-        return y
+dynamic_shapes["w"] = (Dim.AUTO, Dim.DYNAMIC)
+export(model, (w, x, y, z), dynamic_shapes=dynamic_shapes)
 
-dynamic_shapes3 = {
-    "x": {i: Dim(f"inp4_dim{i}") for i in range(inp4.dim())},
-    "y": {i: Dim(f"inp5_dim{i}") for i in range(inp5.dim())},
-}
+######################################################################
+# Static guards also aren't always inherent to the model; they can also come from user-specifications. In fact, a common pitfall leading to shape
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:466]: no need for hyphen in `user-specifications` as a noun, usually you'd use it in a compound adjective

File: intermediate_source/torch_export_tutorial.py [hunk 11]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
-    # suggested fixes below
-    inp4_dim0 = Dim('inp4_dim0', max=16)
-    inp5_dim1 = Dim('inp5_dim1', min=17)
-    inp5_dim0 = inp4_dim1
-    # end of suggested fixes
-    return {
-        "x": {0: inp4_dim0, 1: inp4_dim1},
-        "y": {0: inp5_dim0, 1: inp5_dim1},
-    }
+# Here you might ask why export "specializes"; why we resolve this static/dynamic conflict by going with the static route. The answer is because
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:476]: use `i.e.,` instead of `;` here.

File: intermediate_source/torch_export_tutorial.py [hunk 12]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+# of the symbolic shapes system described above, of symbols and guards. When ``x.shape[0]`` is marked static, we don't allocate a symbol, and compile
+# treating this shape as a concrete integer 4. A symbol is allocated for ``y.shape[1]``, and so we finally emit the guard ``s3 == 4``, leading to
+# specialization.
+#
+# One feature of export is that during tracing, statements like asserts, ``torch._checks()``, and ``if/else`` conditions will also emit guards.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:481]: `_checks()` is not a thing, `_check()` is

File: intermediate_source/torch_export_tutorial.py [hunk 13]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+# Each of these statements emits an additional guard, and the exported program shows the changes; ``s0`` is eliminated in favor of ``s2 + 2``,
+# and ``s2`` now contains lower and upper bounds, reflected in ``range_constraints``.
+#
+# For the if/else condition, you might ask why the True branch was taken, and why it wasn't the ``w.shape[0] != x.shape[0] + 2`` guard that
+# got emitted from tracing. The answer is that export is guided by the sample inputs provided by tracing, and specializes on the branches taken.
+# If different sample input shapes were provided that fail the ``if`` condition, export would trace and emit guards corresponding to the ``else`` branch.
+# Additionally, you might ask why we traced only the ``if`` branch, and if it's possible to maintain control-flow in your program and keep both branches
+# alive. For that, refer to rewriting your model code following the ``Control Flow Ops`` section above.
+#
+# Since we're talking about guards and specializations, it's a good time to talk about the 0/1 specialization issue we brought up earlier.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:520]: Worth structuring this as a subsection or something with an anchor, and referencing (linking) to it when you mentioned it at the beginning?

File: intermediate_source/torch_export_tutorial.py [hunk 14]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+ep = export(
+    torch.nn.Linear(4, 3),
+    (torch.randn(1, 4),),
+    dynamic_shapes={
+        "input": (Dim.AUTO, Dim.STATIC),
+    },
+)
+ep.module()(torch.randn(2, 4))
+
+######################################################################
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:535]: So this is a series of examples of increasing complexity that reads like a novel. Probably good to break it up into subsections with meaningful headers so that readers can pause / resume.

File: intermediate_source/torch_export_tutorial.py [hunk 15]
```diff
@@ -304,237 +304,314 @@ def false_fn(x):
...
+)
+
+######################################################################
+# The expectation with suggested fixes is that the user can interactively copy-paste the changes into their dynamic shapes specification, and successfully export afterwards.
+#
+# Lastly, there's couple nice-to-knows about the options for specification:
+#
+# - ``None`` is a good option for static behavior:
+#   - ``dynamic_shapes=None`` (default) exports with the entire model being static.
+#   - specifying ``None`` at an input-level exports with all tensor dimensions static, and alternatively is also required for non-tensor inputs.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:597]: not "alternatively" but "also"

**Review by avikchaudhuri — APPROVED** [review]
nits

**Review by yushangdi — APPROVED** [review]
LGTM

nt: some weird line breaks here:
Screenshot 2025-01-06 at 10 05 05 AM

File: intermediate_source/torch_export_tutorial.py (hunk 5 above)
Comment (pianpwk) [reply intermediate_source/torch_export_tutorial.py:409]: Ah, this should show up once I figure out how to render the outputs & errors
